### ISSUE TITLE
[Question] Using MQTT with OpenShift Microservices

### ISSUE THREAD

**Issue by WillNilges** [thingsboard/thingsboard#3637]
**Installation**

**Description**
I stood up a ThingsBoard Microservices deployment on OKD 4.5, set up a tenant, and a device. Everything is peachy with HTTP requests as far as I can tell. Doing the following works fine:

```
curl --insecure -v -X POST -d @telemetry.json https://tb-route-http-transport-thingsboard.apps.lab.okd.local/api/v1/<access token>/telemetry --header "Content-Type:application/json"
```
However, while the installation scripts provided an http route, they did not provide an mqtt route. So I added my own.

```
spec:
 host: tb-route-mqtt-transport-thingsboard.apps.lab.okd.local
  path: /api/v1
  to:
    kind: Service
    name: tb-mqtt-transport
    weight: 100
  port:
    targetPort: mqtt
  tls:
    termination: edge
  wildcardPolicy: None
status:
  ingress:
    - host: tb-route-mqtt-transport-thingsboard.apps.lab.okd.local
      routerName: default
      conditions:
        - type: Admitted
          status: 'True'
          lastTransitionTime: '2020-10-23T18:11:43Z'
      wildcardPolicy: None
      routerCanonicalHostname: apps.lab.okd.local
```

And tried using Mosquitto

```
mosquitto_pub -d -h "tb-route-mqtt-transport-thingsboard.apps.lab.okd.local" -t "v1/devices/me/telemetry" -u "$ACCESS_TOKEN" -f "telemetry-data-as-object.json"
```

**However, that failed with an `Error: No route to host` and no other information.**

So I tried `mosquitto_pub` again with `tb-route-node-root-thingsboard.apps.lab.okd.local` since I could use CURL and HTTP with that route (also set up by the scripts). **But that also did not work.**

I also tried performing a Gatling test, and of course, that also did not work.

What am I doing wrong? How do I get this route to work? Do I even need to? Can anybody provide an example of how to get MQTT working on a microservices deployment?

**Environment**

 * OS:  OKD 4.5 on Fedora CoreOS 32 w/ Fedora Server 32 Services VM
 * ThingsBoard: 3.1.1
 * Browser: Firefox 80

**Comment by WillNilges** [comment]
Weather Update:
If I directly portforward to my pod: 
```
[root@okd4-services ~]# oc port-forward tb-mqtt-transport-1 8889:1883
```
then it works.
```
[root@okd4-services test]#  mosquitto_pub -d -h "127.0.0.1" -t "v1/devices/me/telemetry" -u "$ACCESS_TOKEN" -f "telemetry.json" -p 8889 --insecure
Client mosq-GTv66zrfHF7aQGaqjb sending CONNECT
Client mosq-GTv66zrfHF7aQGaqjb received CONNACK (0)
Client mosq-GTv66zrfHF7aQGaqjb sending PUBLISH (d0, q0, r0, m1, 'v1/devices/me/telemetry', ... (199 bytes))
Client mosq-GTv66zrfHF7aQGaqjb sending DISCONNECT
```
But if I use the route
```
[root@okd4-services test]#  mosquitto_pub -d -h "tb-route-mqtt-transport-thingsboard.apps.lab.okd.local" -t "v1/devices/me/telemetry" -u "$ACCESS_TOKEN" -f "telemetry.json" -p 8889 --insecure
Error: Connection refused
```
Nada.

**Comment by WillNilges** [comment]
Another couple of thoughts:

It seems that routes are mostly for HTTP/HTTPS. Do I have to use an Ingress, ExternalIP, or NodePort? Would that have been set up with OKD 3.11?

Can the tb node somehow deal with this?

**Comment by ashmidt-thingsboard** [comment]
You can use nodePort for the mqtt service. For example with a aws we use ELB.
If there is no load balancer we recommend to use nodePort

**Comment by WillNilges** [comment]
@ashmidt-thingsboard Thanks for the reply. I took a look at the NodePort documentation [here](https://docs.openshift.com/container-platform/4.5/networking/configuring_ingress_cluster_traffic/configuring-ingress-cluster-traffic-nodeport.html) and added a NodePort to the mqtt service.

```
oc expose service tb-mqtt-transport --type=NodePort --name=tb-mqtt-ingress --generator="service/v1"
```
This worked. But when I try to use mosquitto to hit the service with the port it opened up,
```
mosquitto_pub -d -h "127.0.0.1" -t "v1/devices/me/telemetry" -u "$ACCESS_TOKEN" -f "telemetry.json" -p 31717 --insecure
```
I get an `Error: Connection refused.` again. I opened the ports on my Service VM firewall. Is there anything else I must do? Perhaps set up a route to the service?

**Comment by WillNilges** [comment]
Also, how would I do this _with_ a load balancer? Does the default thingsboard microservices deployment not have a loadbalancer?

**Comment by WillNilges** [comment]
**TL;DR: Change the type of your tb-mqtt-transport service to NodePort.**

So, after some time, I decided to set this up on raw Kube, and I understand now. Ingresses are primarily used for HTTP requests, and are useful for exposing the web dashboard, the API, the http broker, and whatever else the node layer needs. My ingress looks like this:
```
spec:
  rules:
    - http:
        paths:
          - path: /api/v1/.*
            pathType: ImplementationSpecific
            backend:
              serviceName: tb-http-transport
              servicePort: 8080
          - path: /api
            pathType: ImplementationSpecific
            backend:
              serviceName: tb-node
              servicePort: 8080
          - path: /api/.*
            pathType: ImplementationSpecific
            backend:
              serviceName: tb-node
              servicePort: 8080
          - path: /static/rulenode/.*
            pathType: ImplementationSpecific
            backend:
              serviceName: tb-node
              servicePort: 8080
          - path: /static/.*
            pathType: ImplementationSpecific
            backend:
              serviceName: tb-web-ui
              servicePort: 8080
          - path: /index.html.*
            pathType: ImplementationSpecific
            backend:
              serviceName: tb-web-ui
              servicePort: 8080
          - path: /
            pathType: ImplementationSpecific
            backend:
              serviceName: tb-web-ui
              servicePort: 8080
          - path: /.*
            pathType: ImplementationSpecific
            backend:
              serviceName: tb-node
              servicePort: 8080
```

In kube, at least. Routes handle all that nonsense for you on OpenShift.

So that's all fine and dandy, but to get MQTT working, you need to change the service config, exposing it to the Internet by way of a NodePort. The thing that threw me for a loop was the mystery of the ingress controller. Kube and OpenShift handle this differently, it seems. With Kube, you need to install HAProxy separately. On OpenShift, there are other mechanisms for ingress. But that's beside the point, all you gotta do is change the Service Type to NodePort in the yaml.
``` 
type: NodePort
```

This _should_ work for both Kube and OpenShift.
